app/src/main/java/de/danoeh/antennapod/fragment/preferences/StoragePreferencesFragment.java

```diff
@@ -231,6 +275,24 @@ private void openDirectoryChooser() {
         activity.startActivityForResult(intent, DirectoryChooserActivity.RESULT_CODE_DIR_SELECTED);
     }
 
+    private void openExportPathPicker() {
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT_WATCH) {
+            Intent intentPickAction = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
```

Comment:
What do you think about using `Intent.ACTION_CREATE_DOCUMENT` instead? Then users can directly choose the file name instead of only choosing the folder. Additionally, this removes the scary warning in the following screenshot:

<img width="200" src="https://user-images.githubusercontent.com/5811634/64551848-4ae92980-d336-11e9-90f7-8e85b102034b.png" />


```suggestion
            Intent intentPickAction = new Intent(Intent.ACTION_CREATE_DOCUMENT);
            intentPickAction.addCategory(Intent.CATEGORY_OPENABLE);
            intentPickAction.setType("text/x-opml"); //TODO
            intentPickAction.putExtra(Intent.EXTRA_TITLE, "test.opml"); //TODO
```

This needs a bit more work than the diff I suggested but it's a bit more clean.

Reply:
The whole point of this PR is to let the user choose the export folder. `Intent.ACTION_CREATE_DOCUMENT` lets you specify the file name, but not the output path. I don't see this warning as scary, it makes sense in the context of an application trying to create a new file in a certain storage location. Either way, I've spent a pretty substantial amount of time coding and testing the PR as it currently stands. If the current solution isn't acceptable I can just close this out. I don't have really feel like modifying the PR any longer as it achieves the functionality outlined on the feature request.

Reply:
> Intent.ACTION_CREATE_DOCUMENT lets you specify the file name, but not every available output path

On my device, I can select every location. Are you sure that you can't select every directory? That intent is specifically made for selecting a single file to be saved

Reply:
Did you attempt to save to the document you created? I implemented the following code:

```
    private void openExportPathPickerAlt() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN_MR2) {
            Intent intentPickAction = new Intent(Intent.ACTION_CREATE_DOCUMENT)
                    .addCategory(Intent.CATEGORY_OPENABLE)
                    .setType(CONTENT_TYPE)
                    .putExtra(Intent.EXTRA_TITLE, DEFAULT_OUTPUT_NAME);

            // Creates an implicit intent to launch a file manager which lets
            // the user choose a specific directory to export to.
            try {
                startActivityForResult(intentPickAction, CHOOSE_OPML_EXPORT_PATH);
                return;
            } catch (ActivityNotFoundException e) {
                Log.e(TAG, "No activity found. Should never happen...");
            }
        }

        // If we are using a SDK lower than API 21 or the implicit intent failed
        // fallback to the legacy export process
        export(new OpmlWriter());
    }
```

However when I run this code I receive the following error when attempting to save to certain folders after picking the output file name and path. For example, the downloads folder:

```
09-14 13:49:23.943  8509  8509 E AndroidRuntime: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=65537, result=-1, data=Intent { dat=content://com.android.providers.downloads.documents/document/16 flg=0x43 }} to activity {de.danoeh.antennapod.debug/de.danoeh.antennapod.activity.PreferenceActivity}: java.lang.IllegalArgumentException: Invalid URI: content://com.android.providers.downloads.documents/document/16
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4845)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4886)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7356)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Invalid URI: content://com.android.providers.downloads.documents/document/16
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.provider.DocumentsContract.getTreeDocumentId(DocumentsContract.java:1236)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.support.v4.provider.DocumentFile.fromTreeUri(DocumentFile.java:131)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at de.danoeh.antennapod.asynctask.DocumentFileExportWorker.createExportFile(DocumentFileExportWorker.java:81)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at de.danoeh.antennapod.asynctask.DocumentFileExportWorker.exportObservable(DocumentFileExportWorker.java:41)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at de.danoeh.antennapod.fragment.preferences.StoragePreferencesFragment.export(StoragePreferencesFragment.java:170)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at de.danoeh.antennapod.fragment.preferences.StoragePreferencesFragment.onActivityResult(StoragePreferencesFragment.java:257)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.support.v4.app.FragmentActivity.onActivityResult(FragmentActivity.java:156)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResult(Activity.java:8110)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4838)
09-14 13:49:23.943  8509  8509 E AndroidRuntime: 	... 11 more
```

It appears that in order to use `Intent.ACTION_CREATE_DOCUMENT` you must write a document provider to support each potential/supported storage type:

https://developer.android.com/guide/topics/providers/document-provider.html

Reply:
> It appears that in order to use Intent.ACTION_CREATE_DOCUMENT you must write a document provider to support each potential/supported storage type

Those DocumentProviders are created by other apps you can select files from (for example, DropBox). We do not have to write our own.

> However when I run this code I receive the following error when attempting to save to certain folders after picking the output file name and path. 

`createExportFile` needs to be updated, too. Something like the following:

    return DocumentFile.fromSingleUri(context, outputDirectoryUri);

There is no need to create the file manually anymore. Obviously, the name `outputDirectoryUri` needs to be updated afterwards because it no longer is a directory. Code just for quick explanation, I did not try to build it.

https://developer.android.com/reference/android/support/v4/provider/DocumentFile.html#fromSingleUri(android.content.Context,%20android.net.Uri)

Reply:
I modified the code to use `DocumentFile.fromSingleUri(context, outputFileUri)`, have a look @ByteHamster

Reply:
Looks great now. Thanks a lot for all the review changes. I still haven't tried it but I will do in the next few days. I'm quite confident that it can be merged the way it is.

This won't be in 1.7.3 because I don't want to add any new features just before the release (hoping that we get no error reports and the current RC is the last one). So it might take a bit to get merged (I hope 1-2 weeks).